persistence/soft-index/src/main/java/org/infinispan/persistence/sifs/NonBlockingSoftIndexFileStore.java

```diff
@@ -359,156 +393,138 @@ protected boolean isIndexLoaded() {
-         Thread.currentThread().interrupt();
-         throw log.interruptedWhileStopping(e);
-      } finally {
-         started = false;
-      }
+      }, "soft-index-stop");
    }
 
    @Override
-   public boolean isAvailable() {
-      return new File(configuration.dataLocation()).exists() && new File(configuration.dataLocation()).exists();
+   public CompletionStage<Boolean> isAvailable() {
+      // TODO: does this block?
+      return CompletableFuture.completedFuture(new File(configuration.dataLocation()).exists() &&
+            new File(configuration.dataLocation()).exists());
    }
 
    @Override
-   public synchronized void clear() throws PersistenceException {
-      try {
-         logAppender.clearAndPause();
-         compactor.clearAndPause();
-      } catch (InterruptedException e) {
-         Thread.currentThread().interrupt();
-         throw log.interruptedWhileClearing(e);
-      }
-      try {
-         index.clear();
-      } catch (IOException e) {
-         throw log.cannotClearIndex(e);
-      }
-      try {
-         fileProvider.clear();
-      } catch (IOException e) {
-         throw log.cannotClearData(e);
-      }
-      temporaryTable.clear();
-      compactor.resumeAfterPause();
-      logAppender.resumeAfterPause();
+   public CompletionStage<Void> clear() {
+      return sizeAndClearSequencer.orderOnKey(this, () -> {
+               CompletionStage<CompletionStage<Void>> compactorSubStage = blockingManager.thenApplyBlocking(logAppender.clearAndPause(),
+                     ignore -> compactor.clearAndPause(), "soft-index-clear-compactor");
+               return blockingManager.thenRunBlocking(compactorSubStage.thenCompose(Function.identity()), () -> {
+                  try {
+                     index.clear();
+                  } catch (IOException e) {
+                     throw log.cannotClearIndex(e);
+                  }
+                  try {
+                     fileProvider.clear();
+                  } catch (IOException e) {
+                     throw log.cannotClearData(e);
+                  }
+                  temporaryTable.clear();
+                  compactor.resumeAfterPause();
+               }, "soft-index-clear")
+                     .thenCompose(v -> logAppender.resume());
+            }
+      );
    }
 
    @Override
-   public synchronized int size() {
-      try {
-         logAppender.pause();
-         long size = index.size();
-         return size > Integer.MAX_VALUE ? Integer.MAX_VALUE : (int) size;
-      } catch (InterruptedException e) {
-         Thread.currentThread().interrupt();
-         throw log.sizeCalculationInterrupted(e);
-      } finally {
-         logAppender.resumeAfterPause();
-      }
+   public CompletionStage<Long> size(IntSet segments) {
+      return sizeAndClearSequencer.orderOnKey(this, () ->
+            blockingManager.thenApplyBlocking(logAppender.pause(), ignore -> {
+               try {
+                  return index.size();
```

Comment:
This does not work, now that this store supports segmentation. We can only do this optimization if all the provided segments map to the segments we currently own.